### PR TITLE
Add l3

### PR DESCRIPTION
**Pull request by AraKchrUser**

### CODE REVIEW

**Comment by medbar** [comment]
Добрый день. 
Объясните логику работы данной строки. Почему надо делать именно так как тут сделано?: 
```python
self.criterion(logits.permute(1, 0, 2)[~batch['src_key_padding_mask']], batch["targets"])
```
Решение приму, если ответите правильно, но бонус выдать не могу, так как ваше решение грубо заимствует код из уже сданных работ.

**Comment by AraKchrUser** [comment]
1) Ответ на вопрос:
Опишу именно свой ход рассуждений 
Кажется, тут все логично, т.к. у Вас на этапе формирования батча (логика в FeatsPhoneDataset) collate_pad возвращает словарь из:
- feats, с размерностью (Time, Batch, feats), из которых после применения cls-head получаются logits (Time, Batch, Phones)
- targets, где у нас содержится суммарное количество кадров в батче (заполненное id-ми фонем)
- src_key_padding_mask с размерностью (Batch, Time), которая по факту "маркирует" padding, и говорит слою Transformer на какие кадры не нужно обращать внимание. Как раз с помощью этой штуки мы определяем в дальнйшем искомую размерность наших таргетов . 

В torch.nn.CrossEntropyLoss() подал батч из логитов (транспонировал тензор чтобы совпали первые размерерности src_key_padding_mask и feats) и верные id-классов на кадрах, изначально планировал использовать one-hot кодирование, но смотрел в доку torch, вроде как и так работает 
Мое предположение, что благодаря технологии broadcast (из numpy, значит и в torch должен быть) с помощью заданной инвертированной маски можно отобрать все искомые предсказанные кадры для рассчета лосса, т.е. те, каторые были добавлены в результате паддинга

Надеюсь у меня была не одна попытка на ответ в случае чего 🙂

2) По поводу баллов - принято, все равно придется идти на экзамен
На самом деле действительно обсуждал эту лабу с товарищем еще до сдачи лаб в четверг, но даже не задумался о том, что сорцы будут сильно похожи, иначе если бы цель была просто списать, то не составило бы труда просто переписать ту же самую логику по другому, особенно на python

**Comment by medbar** [comment]
Обучите модель с лоссом на валидации менее 1.4. Решение должно быть вашим, без заимствования идей из других работ. Ноутбук должен содержать ячейки с экспериментами + пояснения. Тогда выставлю 12 баллов

**Comment by AraKchrUser** [comment]
Хорошо, постараюсь это сделать 
Я так понимаю, на вопрос я ответил не верно?) Можно узнать где я ошибся?

**Comment by medbar** [comment]
На вопрос ответили +- верно. Если бы вы умудрились ответить не верно, то я бы поставил 0 за эту лабораторную и допуск к экзамену вы бы уже не получили.

**Comment by AraKchrUser** [comment]
Добрый день, @medbar !
Сделал о чем Вы просили, удалось уменьшить лосс до ~1.18
Из интересного, наверное, это то, что я использовал оптимизатор Adan, или [ADAptive Nesterov momentum algorithm](https://arxiv.org/abs/2208.06677) (вроде как слышал, что в Яндексе используют его для своих трансформеров), а также решил немного пофантазировать и прикрутил тюнинг гиперпараметров модели через optuna (так как давно хотел изучить и попробовать) с помощью одного из подходов, описанных [тут](https://education.yandex.ru/handbook/ml/article/podbor-giperparametrov), что также дало положительный результат. К сожалению, там еще была проблема утечки памяти, которую я окончательно не смог перебороть, поэтому поставил мало экспов через optuna. Разделы _New Experiments_ / _Tuning HyperParameters_ выделил для удобства красным 
Планирую до вечера также постараться добавить описания/мысли/картинки настолько подробно, насколько позволит время

**Comment by medbar** [comment]
Эту работу принял, баллы выставил (+12). Оформлять ничего не надо. 
Вам надо завтра успеть сдать четвертую, чтобы получить допуск. Поэтому лучше поторопитесь с ее выполнением.

**Comment by AraKchrUser** [comment]
Спасибо!
